ar-setup.sh

```diff
@@ -0,0 +1,144 @@
+		exit 1
+	else 
+		echo "Created the artifact repository \"${app_name}\" at the location ${artifact_registry_location}"
+	fi
+else
+	echo "The artifact repository \"${app_name}\" already exists"
+fi
+
+# check whether the service account already exists, and create the service account if it does not exist
+readonly service_account_name="${app_name}-push"
+readonly service_account_email="${service_account_name}@${project}.iam.gserviceaccount.com"
+gcloud iam service-accounts describe "${service_account_email}" >/dev/null 2>&1
+if [ $? -ne 0 ]; then
+	echo "The service account ${service_account_email} does not exist"
+	service_account_description="Service Account for accessing the artifact repository \"${app_name}\""
+	gcloud iam service-accounts create "${service_account_name}" --description "${service_account_description}" >/dev/null 2>&1
+	if [ $? -ne 0 ]; then
+		echo "Failed to create the service account ${service_account_email}"
+		exit 1
+	else
+		echo "Created the service account ${service_account_email}"
+	fi
+else
+	echo "The service account ${service_account_email} already exists"
+fi
+
+# grant the Google service account roles/artifactregistry.writer on the AR repo (this command is idempotent)
+gcloud beta artifacts repositories add-iam-policy-binding "${app_name}" \
+	--location="${artifact_registry_location}" \
+	--member=serviceAccount:"${service_account_email}" \
+	--role=roles/artifactregistry.writer >/dev/null
+if [ $? -ne 0 ]; then
+	echo "Failed to grant the service account ${service_account_email} \"roles/artifactregistry.writer\" on the artifact repository \"${app_name}\""
+	exit 1
+else
+	echo "Granted the service account ${service_account_email} \"roles/artifactregistry.writer\" on the artifact repository \"${app_name}\""
+fi
+
+add_gitlab_project_vars() {
+	# get the gitlab hostname and gitlab project name from app_config_repo
+	readonly https_prefix="https://"
+	readonly https_prefix_len=${#https_prefix}
+	readonly app_config_repo_without_https_prefix=${app_config_repo:https_prefix_len}
+
+	readonly end_of_gitlab_hostname_index=$(expr index ${app_config_repo_without_https_prefix} /)
+	readonly gitlab_hostname_len=https_prefix_len+end_of_gitlab_hostname_index
+	readonly gitlab_hostname=${app_config_repo:0:gitlab_hostname_len}
+
+	gitlab_project=${app_config_repo:gitlab_hostname_len}
+	# replace the slash in the project name with %2F
+	readonly gitlab_project=${gitlab_project//\//%2F}
+
+	# check whether the GitLab project exists
+	readonly output_project_existence=$(curl -s --header "PRIVATE-TOKEN: ${gitlab_access_token}" "${gitlab_hostname}api/v4/projects/${gitlab_project}")
+	readonly not_found_msg="Not Found"
+	if [[ "${output_project_existence}" = *"${not_found_msg}"* ]]; then
+		echo "The GitLab project \"${app_config_repo}\" does not exist. Please run \`appctl init\` to create the project first."
+		exit 1
+	fi
```

Comment:
@janetkuo , just FYI, I added code here to check whether the GitLab project exists or not.

Reply:
Cool!